### ISSUE TITLE
Custom function descriptions

### ISSUE THREAD

**Issue by mnicstruwig** [jackmpcollins/magentic#77]
Hi there,

I'm currently busy using `magentic` and integrating it with parts of `openbb`. Unfortunately, our docstrings are long and comprehensive, which often results in 400 errors when attempting to `@prompt` for the function description being too long.

My attempted workaround was to use a decorator that replaces the docstring with a shortened version, while also preserving the type annotations, but unfortunately this still results in bad function calls that don't conform to the input schema:

```python

# Workaround attempt
def use_custom_docstring(docstring):
    def decorator(func):
        def wrapper(*args, **kwargs):
            return func(*args, **kwargs)
        wrapper.__doc__ = docstring
        wrapper.__annotations__ = func.__annotations__
        return wrapper
    return decorator

```

I suspect this is probably because the signature of the function is inspected directly, but I've not looked too deeply into the `magentic` codebase to understand how the signature is parsed.

A neat solution to this would be to allow users to specify a function description, perhaps as a tuple (or whatever struct you think will work best):

```python
@prompt(
    PROMPT,
    functions=[(my_function, "my function descriotion")],
    model=OpenaiChatModel(model="gpt-4-1106-preview"),
)
def answer_question(world_state: str, question_id: str) -> str:
    ...
```

Thanks for a wonderful project!

**Comment by jackmpcollins** [comment]
Hi @mnicstruwig , `inspect.getdoc` is used internally so changing the `.__doc__` attribute works for the docstring, as in your workaround. For the annotations, if you use the `functools.wraps` decorator then the annotation information can be determined correctly (I guess this must be due to the `.__wrapped__` attribute it adds).

```python
from functools import wraps


def use_custom_docstring(docstring: str):
    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            return func(*args, **kwargs)
        wrapper.__doc__ = docstring
        return wrapper
    return decorator
```

Here's the code I used to check the produced schemas

```python
from magentic.chat_model.function_schema import FunctionCallFunctionSchema

def plus(a: int, b: int) -> int:
    """Original docstring"""
    return a + b

s = FunctionCallFunctionSchema(plus)
print()
print(s.description)
print(s.parameters)

plus_with_new_docstring = use_custom_docstring("new docstring")(plus)

s = FunctionCallFunctionSchema(plus_with_new_docstring)
print()
print(s.description)
print(s.parameters)
```

---

You could achieve a similar API to what you propose by using a regular function rather than a decorator. This would be my preferred solution for the moment - maybe I should include it in magentic as a helper function.

```python
from functools import wraps


def use_custom_docstring(func, docstring: str):
    @wraps(func)
    def wrapper(*args, **kwargs):
        return func(*args, **kwargs)
    wrapper.__doc__ = docstring
    return wrapper

# usage
@prompt(
    PROMPT,
    functions=[use_custom_docstring(my_function, "my function description")],
    model=OpenaiChatModel(model="gpt-4-1106-preview"),
)
def answer_question(world_state: str, question_id: str) -> str:
    ...
```

I hope that works for you. Thanks for using the package!

**Comment by mnicstruwig** [comment]
Hi @jackmpcollins, appreciate the reply over the holiday season! Sorry for replying so late myself.

Thanks for the suggestion! -- I was forgetting the `@wraps` decorator. I ended up still using a decorator pattern, although I don't think it's necessary (as per your example).

A helper function is probably not necessary, given that my use-case is probably more of an edge-case than anything else.

Closing the issue 👍
